### DISCUSSION TITLE
Handling custom scalars with apollo-angular

### DISCUSSION THREAD

**Discussion by pixtim** [the-guild-org/apollo-angular#1688] [Q&A]
# Question
- Is it possible to automatically parse and pack [custom scalars](https://www.apollographql.com/docs/apollo-server/schema/custom-scalars/) with apollo-angular?
- If not, is there a workaround we can use to achieve this? For example, [apollo-link](https://github.com/eturino/apollo-link-scalars) has custom scalar capability.

# Background
I've created a graph that declares custom scalars to represent vectors, matrices and quaternions.
[schema.txt](https://github.com/kamilkisiela/apollo-angular/files/6646578/schema.txt)

My client implementations looks something like this:
```ts
export const partFragment = gql`
  fragment partFields on Part {
    name
    blocks {
      transformation
      cut
      color
    }
  }
`;

@Injectable({ providedIn: "root" })
export class PartService {
  constructor(private apollo: Apollo) {}

  public async getPart(name: string): Promise<Part> {
    interface QueryResult {
      getPart: Part;
    }
    const result = await this.apollo
      .query<QueryResult>({
        query: gql`
          ${partFragment}
          query($name: String!) {
            getPart(name: $name) {
              ...partFields
            }
          }
        `,
        variables: { name },
      })
      .toPromise();
    return result.data.getPart;
  }
```
I'm using a number array to pack the serialized values for each scalar. For example
```gql
{
  "data": {
    "getPart": {
      "name": "testTrivialPart",
      "blocks": [
        {
          "transformation": [ 10, 0, 0, 0, 0, 10, 0, 0, 0, 0, 10, 0, 0, 0, 0, 1 ]
        }
      ]
    }
  }
}
```
The typescript definition for my matrix includes an `elements` array that is populated by a [CustomScalar](https://docs.nestjs.com/graphql/scalars) handler. The client however, only sees the raw number array for the field and does not know how to automatically parse it into a `Matrix4` object. I need to hook in a custom scalar handler to do the mapping automatically when deserializing and serializing.

**Comment by dandean** [comment]
This discussion from apollo-client goes in detail into why this is difficult: https://github.com/apollographql/apollo-client/issues/585
